### PR TITLE
feat: Override `message.new` event function

### LINKED ISSUE #1796: Allow customization of `useMessageNew`
**Motivation**

We can currently only pass `lockChannelOrder` to override the functionality of [useMessageNew](https://github.com/GetStream/stream-chat-react-native/blob/develop/package/src/components/ChannelList/hooks/listeners/useNewMessage.ts) that bumps channels up to the top of the channel list when a new message arrives. However, for some cases (multiple channel lists filtered by channel type for example), it would be helpful to override this functionality in a different way than `lockChannelOrder` does.

**Proposed solution**
Allow passing a function to override the funcitonality of `useMessageNew`

### PR DESCRIPTION
**Pull request by aharwood9**
## 🎯 Goal

Adding the ability to override the `message.new` event with a custom function. Should resolve this - https://github.com/GetStream/stream-chat-react-native/issues/1893#issue-1520763490

fixes #1796 
## 🛠 Implementation details

Added a prop to the `ChannelList` component as per the other existing examples.

## 🎨 UI Changes

<details>
<summary>iOS</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

<details>
<summary>Android</summary>

<table>
    <thead>
        <tr>
            <td>Before</td>
            <td>After</td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>
                
            </td>
            <td>
                
            </td>
        </tr>
    </tbody>
</table>
</details>

## ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Documentation is updated
- [ ] New code is tested in main example apps, including all possible scenarios
  - [ ] SampleApp iOS and Android
  - [ ] Expo iOS and Android

### CODE REVIEW

**Comment by aharwood9** [comment]
hi 👋 Hoping someone can help me with a couple of things for this PR. We are also experiencing the issue reported here - https://github.com/GetStream/stream-chat-react-native/issues/1893#issue-1520763490 and would like to be able to override the functionality when `message.new` is triggered to return early if the wrong channel type. 

Just had a quick question about naming. `onMessageNew` seems like the appropriate name for this function but it is already in use. Do you have any naming suggestions you would prefer? Currently called `onWhateverMessageNew` as a temp solution. 

Also, I know I haven't updated the docs but waiting on the correct naming. Anything else I need to do or any concerns?

**Comment by aharwood9** [comment]
Sorry lastly, I think it is passing but failing as I am not a member of this repo 🤷‍♂️

**Comment by vanGalilea** [comment]
> Sorry lastly, I think it is passing but failing as I am not a member of this repo 🤷‍♂️

Thanks @aharwood9 for your contribution, we really appreciate it. We will run the Sonarcloud scan locally.

**Comment by vanGalilea** [comment]
> @vanGalilea - do you have any suggestions on the naming as per my first comment? Should I call the function something like `onWatchedChannelMessage`? or maybe `onWatchedChannelNewMessage`?

Your approach is a bit different than the one we had in mind. We will revisit your implementation and get back to you with some hints.

And again, we highly appreciate your OSS contribution!

**Comment by aharwood9** [comment]
Sorry for causing chaos but I think I found your proposal in the linked issue. I've updated the PR accordingly. Please let me know if this is still not what you had in mind @vanGalilea

**Comment by vanGalilea** [comment]
Closing in favour of https://github.com/GetStream/stream-chat-react-native/pull/1920

**Comment by khushal87** [comment]
Hey @aharwood9, thanks for raising the PR; I think this is the way to go for now as per https://github.com/GetStream/stream-chat-react-native/issues/1796#issuecomment-1333455599

Please ensure that the branch is updated with the `develop`, `lint`, `test` and `build` phases pass in the `check_pr` action and ask us for review. We will have a look at this.

**Comment by aharwood9** [comment]
> Hey @aharwood9, thanks for raising the PR; I think this is the way to go for now as per [#1796 (comment)](https://github.com/GetStream/stream-chat-react-native/issues/1796#issuecomment-1333455599)
> 
> Please ensure that the branch is updated with the `develop`, `lint`, `test` and `build` phases pass in the `check_pr` action and ask us for review. We will have a look at this.

Thanks, I will work on this today and get the docs updated as well. 🎉

**Comment by aharwood9** [comment]
@khushal87 - This is now ready for review 👍 Tests and Lint are passing, only SonarCloud is failing as I am not a member of this repo. 

I have never used Docusaurus before so wasn't sure how to build locally to ensure everything worked? Couldn't see anything in the ReadMe about it either.

File: docusaurus/docs/reactnative/guides/channel_list_customization.mdx [hunk 1]
```diff
@@ -44,8 +44,8 @@ The default behavior when this event occurs is to query the channel the message
 Thus, if new message appears in some unfrozen channel which current user is member of, it will be added to the list.
 This may not be a desired behavior since the list is only supposed to show frozen channels.
 
-In this case you can replace the default functionality by providing a custom [`onMessageNew`](../core-components/channel_list.mdx#onmessagenew) function as prop to `ChannelList` component.
-`onMessageNew` receives two parameters when called, `setChannels`, a setter function for the internal `channels` state, and `event`, the `Event` object received for the `notification.message_new` event.
+In this case you can replace the default functionality by providing a custom [`OnNewMessageNotification`](../core-components/channel_list.mdx#onnewmessagenotification) function as prop to `ChannelList` component.
+`OnNewMessageNotification` receives two parameters when called, `setChannels`, a setter function for the internal `channels` state, and `event`, the `Event` object received for the `notification.message_new` event.
```
Comment (vanGalilea) [inline docusaurus/docs/reactnative/guides/channel_list_customization.mdx:48]:
```suggestion
`onNewMessageNotification` receives two parameters when called, `setChannels`, a setter function for the internal `channels` state, and `event`, the `Event` object received for the `notification.message_new` event.
```

Comment (vanGalilea) [inline docusaurus/docs/reactnative/guides/channel_list_customization.mdx:48]:
```suggestion
In this case you can replace the default functionality by providing a custom [`onNewMessageNotification`](../core-components/channel_list.mdx#onnewmessagenotification) function as prop to `ChannelList` component.
`onNewMessageNotification` receives two parameters when called, `setChannels`, a setter function for the internal `channels` state, and `event`, the `Event` object received for the `notification.message_new` event.
```

File: docusaurus/docs/reactnative/guides/channel_list_customization.mdx [hunk 2]
```diff
@@ -69,7 +69,7 @@ const customOnMessageNew = async (setChannels, event) => {
   }
 };
 
-<ChannelList filters={filters} onMessageNew={customOnMessageNew} />;
+<ChannelList filters={filters} OnNewMessageNotification={customOnNewMessageNotification} />;
```
Comment (vanGalilea) [inline docusaurus/docs/reactnative/guides/channel_list_customization.mdx:72]:
```suggestion
<ChannelList filters={filters} onNewMessageNotification={customOnNewMessageNotification} />;
```

File: package/src/components/ChannelList/ChannelList.tsx [hunk 3]
```diff
@@ -164,6 +166,34 @@ export type ChannelListProps<
...
+  /**
+   * Override the default listener/handler for event `notification.message_new`
+   * This event is received on channel, which is not being watched.
+   *
+   * @param setChannels Setter for internal state property - `channels`. It's created from useState() hook.
+   * @param event An [Event object](https://getstream.io/chat/docs/event_object) corresponding to `notification.message_new` event
+   *
+   * @overrideType Function
+   * */
+  onNewMessageNotification?: (
```
Comment (vanGalilea) [inline package/src/components/ChannelList/ChannelList.tsx:193]: in terms of naming convention we should add the Suffix event to it just like `onNewMessageEvent`

File: package/src/components/ChannelList/ChannelList.tsx [hunk 4]
```diff
@@ -164,6 +166,34 @@ export type ChannelListProps<
...
+   * Override the default listener/handler for event `message.new`
+   * This event is received on channel, when a new message is added on a channel.
+   *
+   * @param lockChannelOrder If set to true, channels won't dynamically sort by most recent message, defaults to false
+   * @param setChannels Setter for internal state property - `channels`. It's created from useState() hook.
+   * @param event An [Event object](https://getstream.io/chat/docs/event_object) corresponding to `message.new` event
+   *
+   * @overrideType Function
+   * */
+  onNewMessageEvent?: (
```
Comment (vanGalilea) [inline package/src/components/ChannelList/ChannelList.tsx:179]: in terms of naming convention we should remove the Suffix "event" to it just like `onNewMessageNotification `

Comment (vanGalilea) [reply package/src/components/ChannelList/ChannelList.tsx:179]: So this should become `onNewMessage`

File: package/src/components/ChannelList/hooks/listeners/useNewMessageNotification.ts [hunk 5]
```diff
@@ -30,6 +35,8 @@ export const useNewMessageNotification = <
     const handleEvent = async (event: Event<StreamChatGenerics>) => {
       if (typeof onMessageNew === 'function') {
         onMessageNew(setChannels, event);
```
Comment (vanGalilea) [inline package/src/components/ChannelList/hooks/listeners/useNewMessageNotification.ts:37]:
```suggestion
        onMessageNew(setChannels, event);
        console.warn("onMessageNew is deprecated and will be removed in future release. Please use onNewMessageNotification to establish the same behaviour");
```

**Comment by vanGalilea** [comment]
> @khushal87 - This is now ready for review 👍 Tests and Lint are passing, only SonarCloud is failing as I am not a member of this repo.
> 
> I have never used Docusaurus before so wasn't sure how to build locally to ensure everything worked? Couldn't see anything in the ReadMe about it either.

Hey @aharwood9 tnx for your contribution again! I've added my final comments and we can wrap it up ;)

**Comment by aharwood9** [comment]
hi @vanGalilea - sorry to pester but would love to get this merged and released so we can fix the bug we have and release our app? 😃

**Review by vanGalilea — APPROVED** [review]

**Pull request merged**